Read/ReadUntracked/Track for Signal/MaybeSignal/MaybeProp

zakstucke opened the pull request:
Solves #2976. 

In my opinion, given `with()` is allowed and exists, which also has the external user assumption of "I'm just reading the value", but internally it actually does a clone, in 1/3 cases, also applies to `read()` and `track()`. All three cases the user would assume no clone occurs, hence there's no benefit to not having `read()` and `track()` as ergonomic utilities and a standard interface across other signal types.

When a library needs to use `track()` etc, rather than forcing the author to use `Memo` in their external interface, this allows them to use some `Signal` wrapper type, and leaves the performance choice to the external user. 

The `track()` functions are manually implemented rather than using the trait, using the trait would mean `Read` was automatically implemented, but would do a double clone unnecessarily in the derived signal case, hence the manual `Read` implementation instead.

The `WithUntracked` impls are removed, because they're automatically implemented by the new `ReadUntracked` impls

You'll see there's a TODO on the `Read/ReadUntracked` implementations for `MaybeSignal`, I think the Clone could be avoided theoretically for the static case, but interested to hear your thoughts.

zakstucke commented:
Sorry messed up my rebase and git auto closed the PR haha

gbj reviewed (commented):
~~cargo fmt wants this, I think?~~ I am literally unable to use the GitHub interface, that was supposed to be a commit

zakstucke commented:
Haha thanks, don't merge yet, now I know you're happy with the idea, I want to get rid of the clone in MaybeSignal static case, plus find a way to use the Track trait instead of manual impls (maintaining the single clone I'm wanting by currently not using it)

zakstucke commented:
@gbj i've removed my todos and you can merge if you're happy (i didn't implement either)

**MaybeSignal/MaybeProp static case clone**
- I tried adding references to all the traits, but it makes the api so complex, probably not the right idea. 
- Could theoretically `Arc<>` or `StoredValue<>` it, but that would be breaking considering the types are public enums, I'm not sure whether you'd be happy with that.

**manual track impl**
Couldn't simply modify the track trait to work, the manual impls are fine imo

zakstucke commented:
@gbj thinking about this today, would you be open to breaking `MaybeSignal`/`MaybeProp` with `Static(T)` -> `Static(StoredValue<T>)`?

From personal experience, `MaybeSignal`/`MaybeProp` not being `Copy` is not very ergonomic, they're the types that end up being used all over component libraries. 

This benefit could warrant the breaking change and get 2 birds with one stone here: ergonomics of `Copy` and no static clone for `read`/`track` in this PR.

gbj commented:
There's a conversation about this in #1072, which I ultimately closed with the reasoning, "if you want a `Copy` handle to `MaybeSignal<T>` then you should just use `Signal<T>`, because `Signal<T>` implements `From<T>` anyway."

However, getting `T` out of a `Signal<T>` does require cloning it. 

That said, the only real benefit to `MaybeSignal` is if you are separately handling the static and the dynamic halves of the enum, in which case `.read()` doesn't really make sense anyway.

I am wondering if actually deprecating `MaybeSignal` and creating a `Static` branch of `Signal<_>`,  which creates a `StoredValue<T>` and therefore can return a read guard, might actually make the most sense here?

zakstucke commented:
This sounds like a decent idea actually! I'll play around with this. 

Out of interest has anyone benchmarked the cost of `T -> StoredValue<T>`?

zakstucke commented:
@gbj I implemented the static variant of `Signal`.

I think this should actually make `Signal` much more efficient: T's were previously auto morphing into `Signal`'s using `derive()`, meaning clone everywhere on usage rather than no-ops, now `From<T>` doesn't require `Clone` and has clone free reading.

`StoredValue` didn't have a backing `ArcStoredValue` variant, I'm assuming this wasn't done for convenience, but was needed for this so I implemented it. It meant it made sense to extract the `$_value` methods into their own traits too.

**I'm not sure on the method naming for the nonreactive/`Static` `Signal` variant. I've preliminarily gone with `nonreactive()`/`nonreactive_local()`, because static is protected, let me know if something's better!**

In general now `From<T>` doesn't require `Clone`, assuming `T -> StoredValue<T>` is more or less neglible when taking everything else into account, recommending `Signal` for component libs makes sense and as you say maybe even deprecating `MaybeSignal`.

zakstucke commented:
Changed `Signal::nonreactive()` -> `Signal::stored()` (and `stored_local`)

gbj commented on reactive_graph/src/owner/arc_stored_value.rs:
```diff
@@ -0,0 +1,126 @@
+
+/// A reference-counted getter for any value non-reactively.
+///
+/// This is a reference-counted signal, which is `Clone` but not `Copy`.
```
nit: Here and elsewhere (in the new StoredValue traits), these docs refer to these as "signals" but they're really not; I'd probably just replace "signal" with "value".

(Presumably just copy-pasted from the other signal docs)

gbj commented on reactive_graph/src/traits.rs:
```diff
@@ -639,3 +639,199 @@ pub fn panic_getting_disposed_signal(
+
+    /// Updates the value by applying a function, non-reactively, returning the value returned by that function.
+    ///
+    /// # Panics
```
Current behavior in both 0.6 and 0.7 is that `update_value` 1) doesn't return `U` and therefore 2) is guaranteed never to panic -- if the value has already been disposed it simply does nothing. I think we should stick to this pattern.

zakstucke replied:
Yep fair enough, I think I'd copied from StoredValue so the API had changed, but for the set one I was thinking about it as if each disposed error is a downstream user error hence they should see it straight away. But yeah makes sense to avoid an error when possible. Updated

gbj commented on reactive_graph/src/traits.rs:
```diff
@@ -639,3 +639,199 @@ pub fn panic_getting_disposed_signal(
+    /// # Panics
+    /// Panics if you try to access a signal that has been disposed.
+    #[track_caller]
+    fn set_value(&self, value: Self::Value) {
```
`set_value()` definitely doesn't need to panic, here, as it's not even returning a value.

zakstucke replied:
Same reason as `update_value`. Updated

gbj reviewed (commented):
Thanks so much. This is looking really good!

I left a couple minor comments. (In fact looking at the current docs on panicking for 0.7 I think they're just currently wrong for `update_value()` and `set_value`, neither of these needs to panic if they don't return values.

I really appreciate you taking the time, not only to make the PR but to understand things deeply enough to be able to make it!

zakstucke commented:
I think I'm happy with this, I might try a few things afterwards in a separate PR, this is getting too big: 
- sorting out the `Track` trait
- adding `Read`/`Get`... traits for `Option<Signal>`
- deprecating `MaybeSignal` and `MaybeProp` (`MaybeProp` can be deprecated if the traits are added to `Option<Signal>`

gbj commented:
Thanks! Yes, looks good to me for now.
